File: app/src/main/java/dime/android/todo/logic/TaskListAdapter.java
package dime.android.todo.logic;

import android.content.Context;
import android.graphics.Paint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.LinearLayout;
import android.widget.TextView;

import dime.android.todo.R;
import dime.android.todo.ToDo;

public class TaskListAdapter extends BaseAdapter {
    public static final int colors[] = {R.color.low_priority, R.color.normal_priority, R.color.high_priority};

    private ToDo toDoApp;
    private LayoutInflater mInflater;
    private Context context;


    private void setPriorityColor(ViewHolder holder, int priority) {
        holder.priorityColor.setBackgroundResource(colors[priority]);
    }


    private void changeItemStyle(ViewHolder holder, Task task) {
        if (task.isCompleted()) {
            holder.task_name.setPaintFlags(holder.task_name.getPaintFlags() | Paint.STRIKE_THRU_TEXT_FLAG);
            holder.task_name.setTextColor(context.getResources().getColor(R.color.gray));
            holder.list_item_layout.setBackgroundResource(R.color.gray_transparent);
        } else {
            holder.task_name.setPaintFlags(holder.task_name.getPaintFlags() & ~Paint.STRIKE_THRU_TEXT_FLAG);
            holder.task_name.setTextColor(context.getResources().getColor(R.color.white));
            holder.list_item_layout.setBackgroundResource(R.color.transparent);
        }
    }


    public TaskListAdapter(Context context) {
        this.context = context;
        toDoApp = (ToDo) context.getApplicationContext();
        mInflater = LayoutInflater.from(context);
    }


    public int getCount() {
        return toDoApp.taskList.size();
    }


    public Object getItem(int position) {
        return toDoApp.taskList.get(position);
    }


    public long getItemId(int position) {
        return toDoApp.taskList.get(position).getId();
    }


    public View getView(int position, View convertView, ViewGroup parent) {
        ViewHolder holder;

        if (convertView == null) {
            convertView = mInflater.inflate(R.layout.todo_list_item, null);

            holder = new ViewHolder();
            holder.list_item_layout = (LinearLayout) convertView.findViewById(R.id.list_item_layout);
            holder.task_name = (TextView) convertView.findViewById(R.id.task_name);
            holder.priorityColor = (TextView) convertView.findViewById(R.id.priority_color);

            convertView.setTag(holder);
        } else {
            holder = (ViewHolder) convertView.getTag();
        }

		/*
         * Bind the data
		 */
        Task task = toDoApp.taskList.get(position);
        holder.task_name.setText(task.getName());

        // Change the icon and the text style based on the "completed" info
        changeItemStyle(holder, task);

        // Set the correct color (set priority)
        setPriorityColor(holder, task.getPriority());

        return convertView;
    }

    /**
     * The ViewHolder class
     */
    static class ViewHolder {
        LinearLayout list_item_layout;
        TextView task_name;
        TextView priorityColor;
    }

}

File: app/src/main/kotlin/dime/android/todo/main/TaskListAdapter.kt
package dime.android.todo.main

import android.content.Context
import android.graphics.Paint
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.ImageView
import android.widget.TextView
import dime.android.todo.R
import dime.android.todo.db.Task
import dime.android.todo.db.database
import dime.android.todo.extensions.doIfTrue
import dime.android.todo.extensions.inflate
import org.jetbrains.anko.find

/**
 * Created by dime on 06/02/16.
 */

class TaskListAdapter(val context: Context): RecyclerView.Adapter<TaskListAdapter.ViewHolder>() {

    companion object {
        // The priority colors
        val priorityColors = mapOf(
                Task.Priority.LOW to R.color.low_priority,
                Task.Priority.NORMAL to R.color.normal_priority,
                Task.Priority.HIGH to R.color.high_priority)
    }

    // The alpha values of the priorities
    val priorityAlphas = mapOf(Task.Priority.LOW to 0.25f, Task.Priority.NORMAL to 0.5f, Task.Priority.HIGH to 1f)

    // The items
    val tasks = mutableListOf<Task>()

    // The task click listener
    var taskClickListener: ((id: Int) -> Unit)? = null

    // The error delegate (called when there is an error)
    var errorDelegate: ((message: String) -> Unit)? = null

    /**
     * The init block
     */
    init {
        tasks.addAll(context.database.allTasks())
    }

    /**
     * Refreshed the data from the database
     */
    fun refreshDataFromDB() {
        tasks.clear()
        tasks.addAll(context.database.allTasks())
        notifyDataSetChanged()
    }

    /**
     * Called when an item (row) needs to be updated
     */
    override fun onBindViewHolder(vh: ViewHolder, position: Int) {
        // Get the task
        val task = tasks[position]

        // Refresh the row
        vh.dataPosition = position
        vh.taskName.text = task.name
        vh.priorityImage.setColorFilter(vh.itemView.resources.getColor(priorityColors[task.priority]!!))
        refreshUIBasedOnCompleted(vh, task)
    }

    /**
     * Called when an item (row) needs to be created
     */
    override fun onCreateViewHolder(parent: ViewGroup, position: Int): ViewHolder? {
        // Inflate the view
        val view = parent.inflate(R.layout.todo_list_item)
        view.setOnClickListener { taskClickListener?.invoke(tasks[position].id!!) }

        // Create the view holder
        return ViewHolder(view)
    }

    /**
     * Returns the number of items in this adapter
     */
    override fun getItemCount() = tasks.size


    /**
     * Refreshes the UI based on the completed property of the task
     */
    private fun refreshUIBasedOnCompleted(vh: ViewHolder, task: Task) {
        vh.doneLayer.visibility = if (task.completed) View.VISIBLE else View.GONE
        vh.taskName.alpha = if (task.completed) 0.2f else 1.0f
        vh.taskName.paintFlags = if (task.completed) vh.initialPaintFlags or Paint.STRIKE_THRU_TEXT_FLAG else vh.initialPaintFlags
        vh.priorityImage.alpha = if (task.completed) 0.2f else priorityAlphas[task.priority]!!
    }

    /**
     * The view holder
     */
    inner class ViewHolder(val view: View): RecyclerView.ViewHolder(view) {
        val taskName by lazy { view.find<TextView>(R.id.task_name) }
        val foregroundLayer by lazy { view.find<View>(R.id.list_item_layout) }
        val checkBox by lazy { view.find<CheckBox>(R.id.done_checkbox) }
        val priorityImage by lazy { view.find<ImageView>(R.id.priority_image) }
        val doneLayer by lazy { view.find<View>(R.id.done_layer) }

        val initialPaintFlags: Int
        var dataPosition = 0

        init {
            // Setup the UI
            taskName; foregroundLayer; priorityImage
            checkBox.bringToFront()
            doneLayer.alpha = 0.2f
            initialPaintFlags = taskName.paintFlags
            checkBox.setOnCheckedChangeListener { compoundButton, isChecked ->
                val task = tasks[dataPosition]
                task.completed = isChecked

                if (context.database.updateTask(task)) {
                    refreshUIBasedOnCompleted(this, task)
                } else {
                    compoundButton.isChecked = !isChecked
                    task.completed = !isChecked
                    errorDelegate?.invoke(context.getString(R.string.error_while_saving))
                }
            }
        }
    }
}